pymc3/sampling.py

```diff
@@ -373,3 +373,79 @@ def sample_ppc(trace, samples=None, model=None, vars=None, size=None, random_see
+
+    This is a convenience function. NUTS convergence and sampling speed is extremely
+    dependent on the choice of mass/scaling matrix. In our experience, using ADVI
+    to estimate a diagonal covariance matrix and using this as the scaling matrix
+    produces robust results over a wide class of continuous models.
+
+    Parameteres
+    -----------
+    draws : int
+        Number of posterior samples to draw.
+    init : str {'advi', 'advi_map', 'map', 'nuts'}
+        Initialization method to use.
+        * advi : Run ADVI to estimate posterior mean and diagonal covariance matrix.
+        * advi_map: Initialize ADVI with MAP and use MAP as starting point.
+        * map : Use the MAP as starting point.
+        * nuts : Run NUTS and estimate posterior mean and covariance matrix.
+    n_init : int
+        Number of iterations of initializer
+        If 'advi', number of iterations, if 'metropolis', number of draws.
+    sampler : str {'nuts', 'hmc', advi'}
+        Sampler to use. Will be initialized using init algorithm.
+        * nuts : Run NUTS sampler with the init covariance estimation as the scaling matrix.
+        * hmc : Run HamiltonianMC sampler with the init covariance estimation as the scaling matrix.
+        * advi : Sample from variational posterior, requires init='advi'.
+    **kwargs : additional keyword argumemts
+        Additional keyword argumemts are forwared to pymc3.sample()
+
+    Returns
+    -------
+    MultiTrace object with access to sampling values
+    """
+
+    model = pm.modelcontext(model)
+    pm._log.info('Initializing using {}...'.format(init))
+
+    if init == 'advi':
+        v_params = pm.variational.advi(n=n_init)
+        start = pm.variational.sample_vp(v_params, 1)[0]
+        cov = np.power(model.dict_to_array(v_params.stds), 2)
+    elif init == 'advi_map':
+        start = pm.find_MAP()
+        v_params = pm.variational.advi(n=n_init, start=start)
+        cov = np.power(model.dict_to_array(v_params.stds), 2)
+    elif init == 'map':
+        start = pm.find_MAP()
+        cov = pm.find_hessian(point=start)
+
+    elif init == 'nuts':
+        init_trace = pm.sample(step=pm.NUTS(), draws=n_init)
+        cov = pm.trace_cov(init_trace[n_init//2:])
+
+        start = {varname: np.mean(init_trace[varname]) for varname in init_trace.varnames}
+    else:
+        raise NotImplemented('Initializer {} is not supported.'.format(init))
+
+    pm._log.info('Sampling using {}...'.format(sampler))
+    if sampler == 'nuts':
+        step = pm.NUTS(scaling=cov, is_cov=True)
+    elif sampler == 'hmc':
+        step = pm.HamiltonianMC(scaling=cov, is_cov=True)
+    elif sampler != 'advi':
+        raise NotImplemented('Sampler {} is not supported.'.format(init))
+
+    if sampler == 'advi':
+        if init != 'advi':
+            raise ValueError("To sample via ADVI, you have to set init='advi'.")
+        trace = pm.variational.sample_vp(v_params, draws=draws)
+    else:
```

Comment:
should `sampler` instead mimic the `step` argument in `sample`, in that it accepts a function that can sample?  That would elminate some messy string parsing.